### PR TITLE
Cloning (v2): Modular Edition

### LINKED ISSUE #20: Cloning? Only if it can be made 100% modular (Skyrat had it for some time, maybe you can look there).

### PR DESCRIPTION
**Pull request by PeriodicChaos**
## About The Pull Request

_**Cloning (v2): Modular Edition**_  Added back cloning from [monke](https://github.com/Monkestation/MonkeStation/pull/213/files#diff-11e7b481591c36107bd40e6c36d5838f24d9a3979d8e0a3a373b5e45a2679011)!! _(Bee Station Cloning)_

## Why It's Good For The Game

Well, if we want to go a new route we could _disable_ respawns like how **TG** has it other then that its a helpful tool to come back to life and not wait for those pesky crewmates to add more members in the **_Cryopods_** and wait for a **_5-minute timer_** to be up to even start respawning.  All around win for a spaceship game flavor of SS13.

## Notes
Right now cloning works and some Unused/Broken Features are gone I don't think we need much more work done on cloning we can even add it in now if we like the rest can be edited later on all that matters is it works.

## To Do List
- [x] **Make Cloning Work** - _It now lets the body out instead of keeping it for itself very selfish_
- [x] Remove Extra Unused/Broken Features - _Most are removed not all_
- [x] Get a good UI for it - _I did that by updating it to Monke station cloning_
- [ ] Test It - _Gota make sure me changing the code did not just break it all_
cloning
## Changelog
:cl:
add: Cloning (v2): Modular Edition Added Cloning back have fun not waiting the respawn timers and the free points!
/:cl:

Closes #20

### CODE REVIEW

File: tgui/packages/tgui/interfaces/Cloner.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { useBackend } from '../backend';
```
Comment (JohnFulpWillard) [inline tgui/packages/tgui/interfaces/Cloner.js:1]: put this in the modular tgui folder please

File: voidcrew/modules/cloning/code/cloning.dm [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+	if(in_range(user, src) || isobserver(user))
+		. += "<span class='notice'>The status display reads: Cloning speed at <b>[speed_coeff*50]%</b>.<br>Predicted amount of cellular damage: <b>[100-heal_level]%</b><br>"
+		. += "Synthflesh consumption at <b>[round(fleshamnt*90, 1)]cm<sup>3</sup></b> per clone.</span><br>"
+
+		if(efficiency > 5)
+			. += "<span class='notice'>Pod has been upgraded to support autoprocessing and apply beneficial mutations.</span>"
+
+//The return of data disks?? Just for transferring between genetics machine/cloning machine.
+//TO-DO: Make the genetics machine accept them.
+/obj/item/disk/data
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:155]: can disk stuff be in a separate file?

File: voidcrew/modules/cloning/code/objects_datums.dm [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	clonable = FALSE
+
+/datum/brain_trauma/special/beepsky
+	clonable = FALSE
+
+/datum/mood_event/cloned_corpse
+	description = "<span class='boldwarning'>I recently saw my own corpse...</span>\n"
+	mood_change = -6
+
+/datum/quirk/proc/clone_data() //return additional data that should be remembered by cloning
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/objects_datums.dm:39]: This one isnt used anywhere, can we delete it?

File: voidcrew/modules/cloning/code/objects_datums.dm [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+/datum/quirk/proc/on_clone(data) //create the quirk from clone data
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/objects_datums.dm:40]: This also isnt used, since it's on quirk, not trauma.

Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/objects_datums.dm]: I have just found out they are used

File: voidcrew/_DEFINES/cloning_defines.dm [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#define ERROR_COMMITED_SUICIDE 205
+#define ERROR_SOUL_DEPARTED 206
+#define ERROR_SUICIDED_BODY 207
+#define ERROR_SOUL_DAMNED 666
+#define ERROR_UNCLONABLE 901
+
+/// Sent from /obj/machinery/open_machine(): (drop)
+#define COMSIG_MACHINE_OPEN "machine_open"
+/// Sent from /obj/machinery/close_machine(): (atom/movable/target)
+#define COMSIG_MACHINE_CLOSE "machine_close"
```
Comment (JohnFulpWillard) [inline voidcrew/_DEFINES/cloning_defines.dm:32]: When are these signals sent?

File: voidcrew/modules/cloning/code/cloning.dm [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+	var/static/datum/gas_mixture/immutable/cloner/GM //global so that there's only one instance made for all cloning pods
+	if(!GM)
+		GM = new
+	return GM
+
+/obj/machinery/clonepod/proc/get_completion()
+	. = FALSE
+	var/mob/living/mob_occupant = occupant
+	if(mob_occupant)
+		. = (100 * ((mob_occupant.health + 100) / (heal_level + 100)))
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:178]:
```suggestion
	var/mob/living/mob_occupant = occupant
	if(!mob_occupant)
		return FALSE
	return (100 * ((mob_occupant.health + 100) / (heal_level + 100)))
```

File: voidcrew/modules/cloning/code/cloning.dm [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+	var/mob/living/carbon/human/H = new /mob/living/carbon/human(src)
+
+	H.hardset_dna(ui, mutation_index, H.real_name, null, mrace, features)
+
+	if(!HAS_TRAIT(H, TRAIT_RADIMMUNE))//dont apply mutations if the species is Mutation proof.
+		if(efficiency > 2)
+			var/list/unclean_mutations = (GLOB.not_good_mutations|GLOB.bad_mutations)
+			H.dna.remove_mutation_group(unclean_mutations)
+		if(efficiency > 5 && prob(20))
+			H.easy_random_mutate(POSITIVE)
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:228]:
```suggestion
```
We don't want free easy mutations from this

File: voidcrew/modules/cloning/code/cloning.dm [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+
+	if(H)
+		H.faction |= factions
+
+		for(var/V in quirks)
+			var/datum/quirk/Q = new V(H)
+			Q.on_clone(quirks[V])
+
+		for(var/t in traumas)
+			var/datum/brain_trauma/BT = t
```
Comment (JohnFulpWillard) [inline voidcrew/modules/cloning/code/cloning.dm:275]:
```suggestion
		for(var/datum/brain_trauma/BT as anything in traumas)
```

File: voidcrew/_DEFINES/cloning_defines.dm (hunk 5 above)
Comment (PeriodicChaos) [reply voidcrew/_DEFINES/cloning_defines.dm:32]: Im gonna guess when the machine opens and closes

File: voidcrew/modules/cloning/code/cloning.dm [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+	var/radio_key = /obj/item/encryptionkey/headset_med
+	var/radio_channel = RADIO_CHANNEL_MEDICAL
+
+	var/obj/effect/countdown/clonepod/countdown
+
+	var/list/unattached_flesh
+	var/flesh_number = 0
+	var/datum/bank_account/current_insurance
+	fair_market_price = 5 // He nodded, because he knew I was right. Then he swiped his credit card to pay me for arresting him.
+	payment_department = ACCOUNT_MED
```
[the remark this answers is not in the document]
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/cloning.dm:45]: Done so :)

File: voidcrew/modules/cloning/code/cloning.dm [hunk 10]
```diff
@@ -0,0 +1,635 @@
...
+
+/obj/item/disk/data/attack_self(mob/user)
+	read_only0 = !read_only0
+	to_chat(user, "<span class='notice'>You flip the write-protect tab to [read_only0 ? "protected" : "unprotected"].</span>")
+
+/obj/item/disk/data/examine(mob/user)
+	. = ..()
+	. += "The write-protect tab is set to [read_only0 ? "protected" : "unprotected"]."
+
+/obj/item/disk/data/debug
```
[the remark this answers is not in the document]
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/cloning.dm:142]: Yup removing that extra stuff

[the remark this answers is not in the document]
Comment (PeriodicChaos) [reply voidcrew/modules/cloning/code/cloning.dm:142]: Also it seems someone has already added the disk edits the cloning does so nice

**Comment by ZephyrTFA** [comment]
Hi back alive, I'm derg

**Comment by PeriodicChaos** [comment]
Oh... no... AHHHHHHHHHH

**Comment by PeriodicChaos** [comment]
sorry about that it seems i bugged something

**Pull request closed without merging**